File: integrations/deli-j/src/main/java/io/isima/bios/deli/deliverer/DeliveryChannel.java
/*
 * Copyright (C) 2025 Isima, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package io.isima.bios.deli.deliverer;

import io.isima.bios.deli.models.Configuration;
import io.isima.bios.deli.models.InvalidConfigurationException;
import io.isima.bios.models.ImportDestinationConfig;
import io.isima.bios.models.ImportFlowConfig;
import io.isima.bios.sdk.exceptions.BiosClientException;
import java.net.MalformedURLException;
import java.net.URL;
import java.util.Collection;
import java.util.LinkedHashMap;
import java.util.Map;
import java.util.Objects;

public abstract class DeliveryChannel {

  private static Map<String, DeliveryChannel> channels;

  static {
    channels = new LinkedHashMap<>();
  }

  public static DeliveryChannel registerChannel(Configuration configuration,
      ImportDestinationConfig importDestinationConfig) throws InvalidConfigurationException {
    Objects.requireNonNull(configuration);
    Objects.requireNonNull(importDestinationConfig);

    DeliveryChannel channel = getChannel(importDestinationConfig.getImportDestinationId());
    if (channel != null) {
      // REgistered already
      return channel;
    }
    final var destinationType = importDestinationConfig.getType();
    switch (destinationType) {
      case BIOS:
        channel = new BiosDeliveryChannel(configuration, importDestinationConfig);
        break;
      case WEBHOOK:
        channel = new WebhookDeliveryChannel(configuration, importDestinationConfig);
        break;
      default:
        throw new InvalidConfigurationException(
            "importDestination (%s): Destination type %s is unsupported",
            importDestinationConfig.getImportDestinationName(), destinationType);
    }
    channels.put(importDestinationConfig.getImportDestinationId(), channel);
    return channel;
  }

  public static DeliveryChannel getChannel(String channelId) {
    return channels.get(channelId);
  }

  public static Collection<DeliveryChannel> getChannels() {
    return channels.values();
  }

  protected final Configuration configuration;
  protected final ImportDestinationConfig importDestinationConfig;

  protected DeliveryChannel(Configuration configuration,
      ImportDestinationConfig importDestinationConfig) {
    this.configuration = configuration;
    this.importDestinationConfig = importDestinationConfig;
  }

  public String getName() {
    if (importDestinationConfig.getImportDestinationName() == null) {
      return getId();
    }
    return importDestinationConfig.getImportDestinationName();
  }

  public String getId() {
    return importDestinationConfig.getImportDestinationId();
  }

  public ImportDestinationConfig getDestinationConfig() {
    return importDestinationConfig;
  }

  public abstract Deliverer<?> makeDeliverer(ImportFlowConfig importFlowSpec)
      throws InvalidConfigurationException;

  public abstract void start() throws BiosClientException;

  public abstract void stop();

  // Utilities /////////////////////////////////////////////
  protected URL parseUrl(String endpoint, String propertyName)
      throws InvalidConfigurationException {
    try {
      return new URL(endpoint);
    } catch (MalformedURLException e) {
      throw new InvalidConfigurationException("destination '%s': Syntax error in property '%s': %s",
          getName(), propertyName, e.toString());
    }
  }

  protected String retrieveEndpoint(ImportDestinationConfig importDestinationConfig)
      throws InvalidConfigurationException {
    final var endpoint = importDestinationConfig.getEndpoint();
    final var propName = "endpoint";
    if (endpoint == null) {
      throw new InvalidConfigurationException("destination '%s': Property '%s' must be set",
          getName(), propName);
    }
    return endpoint;
  }
}

File: integrations/deli/deli/delivery_channel/delivery_channel.py
#
# Copyright (C) 2025 Isima, Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#
"""delivery_channel: Defines abstract classes used in delivery channels"""
import logging
from abc import ABC, abstractmethod
from configparser import ConfigParser
from typing import List

from ..configuration import Configuration, DataFlowConfiguration, DestinationConfiguration
from ..event_spec import EventSpec


class Deliverer(ABC):
    """Abstract class that takes care of delivering messages.

    Instances are registered to DataFeeder that routs incoming messages to corresponding deliverers.
    Each instance is navigable to corresponding DeliveryChannel that takes care of the physical
    connection.
    """

    @abstractmethod
    def deliver(self, events: List[EventSpec], in_message_user: str, in_message_password: str):
        """Delivers a message.

        Args:
            - events (List[Event]): The incoming message to deliver
            - in_message_user (str): In-message user name or None
            - in_message_password (str): In-message password or NOne
        """


class DeliveryChannel(ABC):
    """Abstract class that takes care of physical connection to an importing data destination.

    This class maintains the physical connection to the target, but does not run actions to deliver.
    Deliverers deliver.
    """

    CHANNEL_BUILDERS = {}
    STUB_CHANNEL = None

    @classmethod
    def build(
        cls, destination_config: DestinationConfiguration, system_config: ConfigParser
    ) -> "DeliveryChannel":
        """Class method to build a delivery channel instance.

        The method creates a delivery channel from specified destination configuration.

        Args:
            destination_config (DestinationConfiguration): Destination configuration. The value may
                be None, in which case a stub channel is created.
        """

        # The stub data forwarder is used for testing
        if destination_config is None:
            return cls.STUB_CHANNEL
        assert isinstance(destination_config, DestinationConfiguration)
        if system_config:
            assert isinstance(system_config, ConfigParser)
        channel_type = destination_config.get_type()
        builder = cls.CHANNEL_BUILDERS.get(channel_type)
        if not builder:
            raise Exception(f"Unsupported channel type: {channel_type}")
        return builder(destination_config, system_config)

    def __init__(self):
        pass

    @abstractmethod
    def setup(self):
        """Sets up the channel.

        Build internal objects and establish necessary connections here.
        """

    @abstractmethod
    def make_deliverer(
        self, configuration: Configuration, flow_config: DataFlowConfiguration
    ) -> Deliverer:
        """Creates a deliverer from specified flow_config and global config.

        Args:
            configuration (Configuration): Deli global config
            flow_config (DataFlowConfiguration): Data flow configuration for the data pipe
        """

    @abstractmethod
    def start(self):
        """Activate delivery actions."""

    @abstractmethod
    def stop(self):
        """Deactivate delivery actions."""

    @abstractmethod
    def shutdown(self):
        """Shuts down the deliverer.

        Terminate resources in the class in this method.
        """


class StubDeliverer(Deliverer):
    """Stub deliverer."""

    def __init__(self):
        self.received_messages = []

    def deliver(self, events: List[EventSpec], in_message_user: str, in_message_password: str):
        logging.info("MESSAGE: %s", events)
        self.received_messages.append(events)


class StubDeliveryChannel(DeliveryChannel):
    """Stub delivery channel."""

    def __init__(self):
        super().__init__()
        self._stub_deliverer = StubDeliverer()

    def setup(self):
        pass

    def make_deliverer(
        self, configuration: Configuration, flow_config: DataFlowConfiguration
    ) -> Deliverer:
        return self._stub_deliverer

    def start(self):
        pass

    def stop(self):
        pass

    def shutdown(self):
        pass

    def get_received_messages(self):
        """Returns received message from deliverer... used for testing."""
        return self._stub_deliverer.received_messages


DeliveryChannel.STUB_CHANNEL = StubDeliveryChannel()
